Add a dialog to modifiy damage roll modifiers

In3luki opened the pull request:
It's only enabled in development  for now and can be opened by holding shift while pressing a damage button.
The UI could still use some refinement but it's looking ok in my opinion.

Weapon Damage:

![Weapon_Normal](https://user-images.githubusercontent.com/41452412/213908990-e24d3574-60da-4300-9cd1-191d21694721.png)

![Weapon_Critical](https://user-images.githubusercontent.com/41452412/213908988-3cfb6a9b-c2ce-4514-b0fb-2c96b0b3e862.png)

![Weapon_Sneak](https://user-images.githubusercontent.com/41452412/213908991-36312aae-73a7-4ac5-8a12-5d789f6d80f9.png)

![Weapon_Deadly](https://user-images.githubusercontent.com/41452412/213908989-2e834799-48a7-40e3-af7b-645ab0547d92.png)

![Splash_Persistent](https://user-images.githubusercontent.com/41452412/213908986-ff3bd21a-3628-4936-b024-77ee9e6fd01e.png)

Spell Damage:

![Spell_Damage](https://user-images.githubusercontent.com/41452412/213908984-ea4a128f-f4df-46b0-8b74-9b972b255b2b.png)

![Spell_Damage2](https://user-images.githubusercontent.com/41452412/213908985-db2945d9-9575-48b6-8a61-f9dc70e6178e.png)

In3luki commented:
Updated to support added spell damage dice.

CarlosFdez commented:
Bothered me at first that spell damage became async but I suppose character strikes work like that too huh?

That said, check modifiers dialog is displayed in CheckPF2e, but this is displayed when computing formulas. That bothers me a little bit but I can imagine why it'd be a major pain in the rear to convert all that, especially considering how inconsistent instance assignment is.

CarlosFdez commented:
The label seems to have superflous text. Not sure if you need +2d6 precision if another column already lists it.

In3luki commented:
> Bothered me at first that spell damage became async but I suppose character strikes work like that too huh?
> 
> That said, check modifiers dialog is displayed in CheckPF2e, but this is displayed when computing formulas. That bothers me a little bit but I can imagine why it'd be a major pain in the rear to convert all that, especially considering how inconsistent instance assignment is.

I wanted to keep it a getter but there is no other way to intercept the modifiers before the forumla is created.

> The label seems to have superflous text. Not sure if you need +2d6 precision if another column already lists it.

The labels are created somewhere else. I can look into stripping the damage dice from it if you want.

CarlosFdez commented:
Ah, the labels are a symptom of our awkward breakdown creation pipeline. Don't worry about it then.

As for making spell damage a method, we need to do that anyways for outcomes.

In3luki commented:
Now uses `eventToRollParams` which makes it adhere to the check roll show dialog setting.

CarlosFdez commented on src/module/actor/character/document.ts:
```diff
@@ -1815,32 +1816,37 @@ class CharacterPF2e extends CreaturePF2e {
+                    domains,
+                };
+
+                const rollParams = params.event ? eventToRollParams(params.event) : {};
```
Would there be any consequences here if eventToRollParams() allowed event to be optional, and purely used the settings in that situation?

In3luki replied:
Probably not. I'll check it out.

CarlosFdez replied:
I could imagine you could ... it into the damage context, then you'd only need to tweak skipDialog specifically.

CarlosFdez commented:
Would it be cleaner to create the DamageRollContext in spell's getFormula() then return it instead of returning modifiers/damage dice/etc? Those returns exist to create the damage context in rollDamage().

In3luki commented:
Yeah, spell damage preparation is a bit all over the place. I can take another look at it.

CarlosFdez commented on src/module/system/damage/weapon.ts:
```diff
@@ -27,16 +27,24 @@ import { DegreeOfSuccessIndex, DEGREE_OF_SUCCESS } from "@system/degree-of-succe
-    ): WeaponDamageTemplate | null {
+        options: Set<string> = new Set(),
+        context: DamageRollContext
+    ): Promise<WeaponDamageTemplate | null> {
```
If you're adding new params, might as well convert it to an object like WeaponDamagePF2e#calculate(). That said, isn't options part of context?

In3luki replied:
I'm not sure what the intention was there anymore. The options are indeed part of the context.

In3luki commented:
- Made `eventToRollParams` event parameter optional.
- Moved `DamageRollContext` and `SpellDamageTemplate` creation to `SpellPF2e#getDamage`.
- Changed `WeaponDamagePF2e.calculateStrikeNPC` params to an object.

CarlosFdez commented on src/module/item/spell/document.ts:
```diff
@@ -826,34 +867,13 @@ class SpellPF2e extends ItemPF2e {
         }
 
-        const spellDamage = this.damage;
+        const rollParams = eventToRollParams(event);
+        const spellDamage = await this.getDamage({ ...rollParams });
```
```suggestion
        const spellDamage = await this.getDamage(eventToRollParams(event));
```

CarlosFdez commented:
this: Embedded<SpellPF2e> is working fine? I remember removing those because it caused nonesense with subclassing.

CarlosFdez commented on src/module/actor/helpers.ts:
```diff
@@ -337,22 +338,31 @@ function strikeFromMeleeItem(item: Embedded<MeleePF2e>): NPCStrike {
+                domains,
+            };
+
+            const rollParams = params.event ? eventToRollParams(params.event) : {};
```
Missed a spot :^)

CarlosFdez commented on src/module/item/spell/document.ts:
```diff
@@ -324,6 +349,8 @@ class SpellPF2e extends ItemPF2e {
             // Add damage dice to instances
             for (const dice of damageDice) {
                 if (dice.override) continue;
+                // Skip dice that were ignored in the dialog
+                if (dice.ignored) continue;
```
```suggestion
                if (dice.override || dice.ignored) continue;
```

CarlosFdez replied:
Well, this is likely to change anyways when I move all of this to the damage pipeline. Don't worry about it.

In3luki commented:
Updated
